sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,459 @@
+import math
+import warnings
+
+import numpy as np
+import dpctl
+from sklearn.exceptions import DataConversionWarning
+
+from sklearn_numba_dpex.utils._device import _DeviceParams
+
+from sklearn_numba_dpex.kmeans.kernels import (
+    make_initialize_to_zeros_1dim_kernel,
+    make_initialize_to_zeros_2dim_kernel,
+    make_initialize_to_zeros_3dim_kernel,
+    make_copyto_kernel,
+    make_broadcast_division_kernel,
+    make_center_shift_kernel,
+    make_half_l2_norm_dim0_kernel,
+    make_sum_reduction_1dim_kernel,
+    make_sum_reduction_2dim_kernel,
+    make_sum_reduction_3dim_kernel,
+    make_fused_fixed_window_kernel,
+    make_assignment_fixed_window_kernel,
+)
+
+
+def _check_power_of_2(e):
+    if e != 2 ** (math.log2(e)):
+        raise ValueError(f"Expected a power of 2, got {e}")
+    return e
+
+
+class LLoydKMeansDriver:
+    """GPU optimized implementation of Lloyd's k-means.
+
+    Inspired by the "Fused Fixed" strategy exposed in:
+    Kruliš, M., & Kratochvíl, M. (2020, August). Detailed analysis and
+    optimization of CUDA K-means algorithm. In 49th International
+    Conference on Parallel Processing-ICPP (pp. 1-11).
+
+    The current implementation is called "fused fixed", it consists in a sliding window
+    of fixed size on the value of the centroids that work items use to accumulate the
+    distances of a given sample to each centroids. It is followed, in a same kernel, by
+    an update of new centroids in a context of memory privatization to avoid a too high
+    cost of atomic operations.
+
+    This class instantiates into a callable that mimics the interface of sklearn's
+    private function `_kmeans_single_lloyd` .
+
+    Parameters
+    ----------
+    preferred_work_group_size_multiple : int
+        The kernels will use this value to optimize the distribution of work items. If
+        None, it is automatically fetched with pyopencl if possible, else a default of
+        64 is applied. It is required to be a power of two.
```

Comment:
Really small nit:
```suggestion
        64 is applied. It must be a power of two.
```

Reply:
It must more correct here ? i'd think is required to is more correct since being a power of two is a requirement.